src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -390,8 +457,35 @@ public IEnumerable<MethodWithGCInfo> EnumerateCompiledMethods()
             }
         }
 
-        private readonly Dictionary<ReadyToRunFixupKind, Dictionary<TypeAndMethod, MethodFixupSignature>> _methodSignatures =
-            new Dictionary<ReadyToRunFixupKind, Dictionary<TypeAndMethod, MethodFixupSignature>>();
+        private class MethodFixupKey : IEquatable<MethodFixupKey>
```

Comment:
>MethodFixupKey [](start = 22, length = 14)

Thinking a bit more about MethodFixupKey and TypeFixupKey, would it be easier here to just use Tuple`2?

Reply:
`Tuple'2` is a class - these should be valuetypes so that we don't allocate every time we want to do a lookup.

There's ValueTuple but it uses the default comparer to compare equality which tends to be slower. Also `GetHashCode` in `ValueTuples` has handling for object being null, etc. I've never measured if this matters, but NodeFactory feels like something that is used pretty often for the explicit structs to be worth the extra "boring struct" (especially because we want to use the same thing everywhere for consistency).

Reply:
In my previous Crossgen2 work I was mostly doing what Michal suggests - creating boring structs representing the various keys with explicit implementation of IEquatable - basically for the reasons he mentioned (not mentioning the fact that he code reviewed 80% of my changes so he naturally kept steering me towards what he perceived as appropriate and performant).